File: src/ui/render/highlight.ts
/**
 * Dependency-free syntax highlighting for fenced code blocks.
 *
 * Security invariant: the tokenizer only ever *slices* the source, and every
 * slice is passed through `escapeHtml` before it reaches the output. The class
 * names are literals from this file, never source-derived, so no path here can
 * emit markup a document supplied — a fence containing `</code><script>` comes
 * back as text.
 *
 * Speed invariant: one master regex per language, built from patterns that
 * cannot backtrack and that always CONSUME an unterminated opener rather than
 * retrying it at every later position (see highlight-langs.ts). Both halves
 * are load-bearing: without the first, 600KB of `"\` pairs blocked the event
 * loop for five minutes; without the second, 900KB of `/*a` blocked it for
 * two. The size cap below is a backstop, not the mechanism.
 */
import { resolveLang, SPECS, type Spec, type Token } from "./highlight-langs.js";
import { diffTokens, mdTokens, yamlTokens } from "./highlight-lines.js";
import { escapeHtml } from "./markdown-inline.js";

/** A backstop past which highlighting stops earning its cost; text still renders. */
const MAX_CHARS = 1_000_000;

const NUM = "\\b\\d[\\w.]*";
const IDENT = "[A-Za-z_$@][\\w$]*";
const PUNCT = "[{}()\\[\\].,;:+\\-*/%=<>!&|?~^#\\\\]+";

const compiled = new Map<string, RegExp>();

function master(name: string, spec: Spec): RegExp {
	const cached = compiled.get(name);
	if (cached !== undefined) return cached;
	const re = new RegExp(`(${spec.comment})|(${spec.string})|(${NUM})|(${IDENT})|(${PUNCT})`, "g");
	compiled.set(name, re);
	return re;
}

/** What follows the match, bounded — used for "is this a call?" and JSON keys. */
function peek(code: string, at: number): string {
	return code.slice(at, at + 8);
}

function classOf(m: RegExpExecArray, code: string, end: number, spec: Spec): string {
	if (m[1] !== undefined) return "hl-com";
	if (m[2] !== undefined)
		return spec.key === true && /^\s*:/.test(peek(code, end)) ? "hl-fn" : "hl-str";
	if (m[3] !== undefined) return "hl-num";
	if (m[4] !== undefined) {
		const word = m[4];
		if (spec.keywords.has(word)) return "hl-kw";
		if (word.startsWith("$") || word.startsWith("@")) return "hl-fn";
		return /^\s*\(/.test(peek(code, end)) ? "hl-fn" : "";
	}
	return "hl-punct";
}

function codeTokens(code: string, name: string, spec: Spec): Token[] {
	const re = master(name, spec);
	re.lastIndex = 0;
	const out: Token[] = [];
	let last = 0;
	for (let m = re.exec(code); m !== null; m = re.exec(code)) {
		if (m[0] === "") {
			re.lastIndex++; // a zero-width match would spin forever
			continue;
		}
		if (m.index > last) out.push({ cls: "", text: code.slice(last, m.index) });
		out.push({ cls: classOf(m, code, re.lastIndex, spec), text: m[0] });
		last = re.lastIndex;
	}
	if (last < code.length) out.push({ cls: "", text: code.slice(last) });
	return out;
}

function tokensFor(code: string, name: string): Token[] {
	if (name === "diff") return diffTokens(code);
	if (name === "yaml") return yamlTokens(code);
	if (name === "md") return mdTokens(code);
	const spec = SPECS[name];
	return spec === undefined ? [{ cls: "", text: code }] : codeTokens(code, name, spec);
}

/**
 * Escaped HTML for `code`, with `<span class="hl-*">` around what it knows.
 * An unsupported language (or anything that goes wrong) returns the escaped
 * source unchanged — this function never throws.
 */
export function highlightCode(code: string, lang: string): string {
	if (typeof code !== "string" || code === "") return "";
	const name = resolveLang(lang);
	if (name === "" || code.length > MAX_CHARS) return escapeHtml(code);
	try {
		let html = "";
		for (const t of tokensFor(code, name)) {
			const text = escapeHtml(t.text);
			html += t.cls === "" ? text : `<span class="${t.cls}">${text}</span>`;
		}
		return html;
	} catch {
		return escapeHtml(code);
	}
}

File: crates/flusk-core/src/render/highlight.rs
//! Escaped HTML for `code`, with `<span class="hl-*">` around what the
//! tokenizer knows. The reference's header documents 600KB inputs blocking
//! the event loop for five minutes; this port removes that class of bug by
//! being fast, not by changing output — including the 1M-unit backstop.
//!
//! Security invariant, unchanged: the tokenizer only ever *slices* the
//! source, every slice passes through `escape_html` before it reaches the
//! output, and the class names are literals — a fence containing
//! `</code><script>` comes back as text.

use super::escape::escape_html;
use super::hl_langs::{is_keyword, resolve_lang, spec, Family, Spec};
use super::hl_lines::diff_tokens;
use super::hl_md::md_tokens;
use super::hl_scan::{Scanner, Token};
use super::hl_yaml::yaml_tokens;
use super::js::utf16_len;

/// A backstop past which highlighting stops earning its cost; text renders.
const MAX_CHARS: usize = 1_000_000;

fn class_of(s: &Scanner, group: u8, start: usize, end: usize, family: Family, sp: &Spec) -> &'static str {
    match group {
        1 => "hl-com",
        2 => {
            if sp.key && s.peek_ws_then(end, ':') {
                "hl-fn" // a JSON object key
            } else {
                "hl-str"
            }
        }
        3 => "hl-num",
        4 => {
            let word = s.slice(start, end);
            if is_keyword(family, word) {
                "hl-kw"
            } else if word.starts_with('$') || word.starts_with('@') || s.peek_ws_then(end, '(') {
                "hl-fn" // a sigiled name, or what looks like a call
            } else {
                ""
            }
        }
        _ => "hl-punct",
    }
}

fn code_tokens<'a>(code: &'a str, family: Family, sp: &Spec) -> Vec<Token<'a>> {
    let s = Scanner::new(code);
    let n = s.len();
    let mut out = Vec::new();
    let mut last = 0;
    let mut p = 0;
    while p < n {
        let Some((group, end)) = s.next_match(sp, p) else {
            p += 1; // bump-along: no alternative matched here
            continue;
        };
        if p > last {
            out.push(Token { cls: "", text: s.slice(last, p) });
        }
        out.push(Token { cls: class_of(&s, group, p, end, family, sp), text: s.slice(p, end) });
        last = end;
        p = end;
    }
    if last < n {
        out.push(Token { cls: "", text: s.slice(last, n) });
    }
    out
}

fn tokens_for(code: &str, family: Family) -> Vec<Token<'_>> {
    match family {
        Family::Diff => diff_tokens(code),
        Family::Yaml => yaml_tokens(code),
        Family::Md => md_tokens(code),
        _ => match spec(family) {
            Some(sp) => code_tokens(code, family, &sp),
            None => vec![Token { cls: "", text: code }],
        },
    }
}

/// Escaped HTML for `code`, spans around what it knows. An unsupported
/// language returns the escaped source unchanged; this function never fails.
pub fn highlight_code(code: &str, lang: &str) -> String {
    if code.is_empty() {
        return String::new();
    }
    let Some(family) = resolve_lang(lang) else { return escape_html(code) };
    // The cap counts UTF-16 units, as `code.length` does; counting is only
    // needed at all once the (always larger) byte length crosses it.
    if code.len() > MAX_CHARS && utf16_len(code) > MAX_CHARS {
        return escape_html(code);
    }
    let mut html = String::with_capacity(code.len());
    for t in tokens_for(code, family) {
        let text = escape_html(t.text);
        if t.cls.is_empty() {
            html.push_str(&text);
        } else {
            html.push_str("<span class=\"");
            html.push_str(t.cls);
            html.push_str("\">");
            html.push_str(&text);
            html.push_str("</span>");
        }
    }
    html
}
